lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetSchemaConverter.java

```diff
@@ -102,25 +116,23 @@ private org.apache.parquet.schema.Type getPrimitiveType(Type type, String name,
         }
         if (type instanceof DecimalType) {
             DecimalType decimalType = (DecimalType) type;
-            if (decimalType.getPrecision() <= 9) {
-                return Types.optional(PrimitiveType.PrimitiveTypeName.INT32)
-                        .as(OriginalType.DECIMAL)
-                        .precision(decimalType.getPrecision())
-                        .scale(decimalType.getScale()).named(name);
-            }
-            else if (decimalType.isShort()) {
-                return Types.optional(PrimitiveType.PrimitiveTypeName.INT64)
-                        .as(OriginalType.DECIMAL)
-                        .precision(decimalType.getPrecision())
-                        .scale(decimalType.getScale()).named(name);
-            }
-            else {
-                return Types.optional(PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY)
-                        .length(16)
-                        .as(OriginalType.DECIMAL)
-                        .precision(decimalType.getPrecision())
-                        .scale(decimalType.getScale()).named(name);
+            // Apache Hive version 3 or lower does not support reading decimals encoded as INT32/INT64
+            if (!useLegacyDecimalEncoding) {
+                if (decimalType.getPrecision() <= 9) {
+                    return Types.optional(PrimitiveType.PrimitiveTypeName.INT32)
+                            .as(decimalType(decimalType.getScale(), decimalType.getPrecision()))
+                            .named(name);
+                }
+                if (decimalType.isShort()) {
+                    return Types.optional(PrimitiveType.PrimitiveTypeName.INT64)
+                            .as(decimalType(decimalType.getScale(), decimalType.getPrecision()))
+                            .named(name);
+                }
             }
+            return Types.optional(PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY)
+                    .length(PRECISION_TO_BYTE_COUNT[decimalType.getPrecision() - 1])
```

Comment:
we used to have length(16).

should the behavior depend on useLegacyDecimalEncoding ?

Reply:
The previous logic of using 16 bytes for all long decimals was wasting space. It's better that we base it on the bytes required for a given precision.
It's also consistent with Hive's logic https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/convert/HiveSchemaConverter.java#L126

Reply:
Right, it does end up improving behaviour for long decimals in delta, I'll extract it in another commit.